File: sycl/sycl-rs-sys/src/queue.cpp
//
// Copyright (C) 2026 Intel Corporation
//
// Under the MIT License or the Apache License v2.0.
// See LICENSE-MIT and LICENSE-APACHE for license information.
// SPDX-License-Identifier: MIT OR Apache-2.0
//

#include "sycl-rs-sys/include/queue.hpp"
#include "sycl-rs-sys/include/utils.hpp"
#include "sycl-rs-sys/src/queue-sys.rs.h"

using sycl::ext::intel::property::queue::immediate_command_list;
using sycl::property::queue::in_order;
using sycl_shims::utils::vec_to_vector;

namespace syclexp = sycl::ext::oneapi::experimental;

namespace sycl_shims::queue {
std::unique_ptr<Queue> new_queue() {
  return std::make_unique<Queue>(sycl::queue({in_order()}));
}

std::unique_ptr<Queue> new_queue_immediate() {
  return std::make_unique<Queue>(
      sycl::queue({in_order(), immediate_command_list()}));
}

std::unique_ptr<Queue> new_queue_from_device(Device const &device) {
  return std::make_unique<Queue>(sycl::queue(device, {in_order()}));
}

std::unique_ptr<Context> get_context(Queue const &queue) {
  return std::make_unique<Context>(queue.get_context());
}

std::unique_ptr<Queue> clone(Queue const &queue) {
  return std::make_unique<Queue>(sycl::queue(queue));
}

std::unique_ptr<Event> memset(std::unique_ptr<Queue> &queue, std::uint8_t *ptr,
                              int value, std::size_t num_bytes,
                              rust::Vec<EventPtr> dep_events) {
  return std::make_unique<Event>(queue->memset(
      ptr, value, num_bytes, vec_to_vector(std::move(dep_events))));
}

std::unique_ptr<Event> barrier(std::unique_ptr<Queue> &queue,
                               rust::Vec<EventPtr> dep_events) {
  return std::make_unique<Event>(
      queue->ext_oneapi_submit_barrier(vec_to_vector(std::move(dep_events))));
}

void wait(std::unique_ptr<Queue> &queue) { queue->wait(); }

template <int Dimensions>
std::unique_ptr<Event>
launch(std::unique_ptr<Queue> &queue, sycl::nd_range<Dimensions> nd_range,
       Kernel const &kernel,
       rust::Slice<rust::slice<std::uint8_t const> const> args) {
  return std::make_unique<Event>(queue->submit([&](sycl::handler &cgh) {
    for (std::size_t i = 0; i < args.size(); ++i)
      cgh.set_arg(i, syclexp::raw_kernel_arg(args[i].data(), args[i].size()));

    cgh.parallel_for(nd_range, kernel);
  }));
}

std::unique_ptr<Event>
launch_1d(std::unique_ptr<Queue> &queue, Range1 global_size, Range1 local_size,
          Kernel const &kernel,
          rust::Slice<rust::slice<std::uint8_t const> const> args) {
  return launch(queue,
                sycl::nd_range<1>{{global_size.data[0]}, {local_size.data[0]}},
                kernel, args);
}

std::unique_ptr<Event>
launch_2d(std::unique_ptr<Queue> &queue, Range2 global_size, Range2 local_size,
          Kernel const &kernel,
          rust::Slice<rust::slice<std::uint8_t const> const> args) {
  return launch(queue,
                sycl::nd_range<2>{{global_size.data[0], global_size.data[1]},
                                  {local_size.data[0], local_size.data[1]}},
                kernel, args);
}

std::unique_ptr<Event>
launch_3d(std::unique_ptr<Queue> &queue, Range3 global_size, Range3 local_size,
          Kernel const &kernel,
          rust::Slice<rust::slice<std::uint8_t const> const> args) {
  return launch(
      queue,
      sycl::nd_range<3>{
          {global_size.data[0], global_size.data[1], global_size.data[2]},
          {local_size.data[0], local_size.data[1], local_size.data[2]}},
      kernel, args);
}

std::unique_ptr<Event> memcpy(std::unique_ptr<Queue> &queue, std::uint8_t *dest,
                              std::uint8_t const *src, std::size_t num_bytes,
                              rust::Vec<EventPtr> dep_events) {
  return std::make_unique<Event>(queue->memcpy(
      dest, src, num_bytes, vec_to_vector(std::move(dep_events))));
}
} // namespace sycl_shims::queue

File: sycl/sycl-rs/src/queue.rs
//
// Copyright (C) 2026 Intel Corporation
//
// Under the MIT License or the Apache License v2.0.
// See LICENSE-MIT and LICENSE-APACHE for license information.
// SPDX-License-Identifier: MIT OR Apache-2.0
//

use crate::Result;
use bytemuck::Pod;
use sycl_rs_sys::{queue::ffi, types::ffi::EventPtr};

use crate::{
    buffer::{
        Buffer, DeviceBuffer, EnqueuedBuffer, EnqueuedDeviceBuffer, EnqueuedHostBuffer,
        EnqueuedSharedBuffer, HostBuffer, SharedBuffer,
    },
    context::Context,
    device::Device,
    event::Event,
    kernel::{Kernel, KernelArgumentList},
    range::{NdRange, ValidDimension},
    usm::{UsmAlloc, UsmAllocator},
};

/// The `Queue` connects a host program to a single device. Programs submit tasks to a device via the
/// `Queue` and may monitor the `Queue` for completion. A program initiates the task by submitting
/// a kernel.
pub struct Queue(pub(crate) cxx::UniquePtr<ffi::Queue>);

impl Queue {
    /// Construct a `Queue` based on the device returned from the default selector.
    pub fn new() -> Self {
        Self(ffi::new_queue())
    }

    /// Construct an immediate `Queue` based on the device returned from the default selector.
    pub fn new_immediate() -> Self {
        Self(ffi::new_queue_immediate())
    }

    /// Returns the SYCL queue’s context.
    pub fn get_context(&self) -> Context {
        ffi::get_context(&self.0).into()
    }

    /// Allocates zeroed memory and creates a host-side [`Buffer`] that can store an array of T.
    pub fn alloc_host<T: Pod>(&mut self, len: usize) -> Result<EnqueuedHostBuffer<T>> {
        unsafe {
            let mut buffer = self.alloc_uninit_host(len);
            self.memset(&mut buffer, 0)
                .map(|event| EnqueuedBuffer::new(buffer, event))
        }
    }

    /// Allocates zeroed memory and creates a shared [`Buffer`] that can store an array of T.
    pub fn alloc_shared<T: Pod>(&mut self, len: usize) -> Result<EnqueuedSharedBuffer<T>> {
        unsafe {
            let mut buffer = self.alloc_uninit_shared(len);
            self.memset(&mut buffer, 0)
                .map(|event| EnqueuedBuffer::new(buffer, event))
        }
    }

    /// Allocates zeroed memory and creates a device [`Buffer`] that can store an array of T.
    pub fn alloc_device<T: Pod>(&mut self, len: usize) -> Result<EnqueuedDeviceBuffer<T>> {
        unsafe {
            let mut buffer = self.alloc_uninit_device(len);
            self.memset(&mut buffer, 0)
                .map(|event| EnqueuedBuffer::new(buffer, event))
        }
    }

    /// Allocates memory and creates a host-side [`Buffer`] that can store an array of T.
    /// Safety: the buffer contents are uninitialized.
    pub unsafe fn alloc_uninit_host<T>(&self, len: usize) -> HostBuffer<T> {
        let allocator = UsmAllocator::from(self);
        unsafe { Buffer::new(allocator, len) }
    }

    /// Allocates memory and creates a shared [`Buffer`] that can store an array of T.
    /// Safety: the buffer contents are uninitialized.
    pub unsafe fn alloc_uninit_shared<T>(&self, len: usize) -> SharedBuffer<T> {
        let allocator = UsmAllocator::from(self);
        unsafe { Buffer::new(allocator, len) }
    }

    /// Allocates memory and creates a device-side [`Buffer`] that can store an array of T.
    /// Safety: the buffer contents are uninitialized.
    pub unsafe fn alloc_uninit_device<T>(&self, len: usize) -> DeviceBuffer<T> {
        let allocator = UsmAllocator::from(self);
        unsafe { Buffer::new(allocator, len) }
    }

    /// Sets memory allocated with USM allocations.
    /// Safety: the caller must make sure the underlying memory isn't being aliased somewhere else.
    pub unsafe fn memset<T, A: UsmAlloc>(
        &mut self,
        buffer: &mut Buffer<T, A>,
        value: i32,
    ) -> Result<Event> {
        unsafe { self.memset_with_deps(buffer, value, &[]) }
    }

    /// Sets memory allocated with USM allocations after all specified events finish.
    /// Safety: the caller must make sure the underlying memory isn't being aliased somewhere else.
    pub unsafe fn memset_with_deps<T, A: UsmAlloc>(
        &mut self,
        buffer: &mut Buffer<T, A>,
        value: i32,
        dep_events: &[&Event],
    ) -> Result<Event> {
        let ptr = buffer.get_byte_ptr();
        let num_bytes = buffer.get_byte_size();
        let dep_events = dep_events
            .iter()
            .map(|e| EventPtr {
                ptr: (*e).clone().0,
            })
            .collect::<Vec<_>>();
        unsafe { ffi::memset(&mut self.0, ptr, value, num_bytes, dep_events) }.map(Into::into)
    }

    /// Submits a barrier to the queue.
    pub fn barrier(&mut self) -> Result<Event> {
        self.barrier_with_deps(&[]).map(Into::into)
    }

    /// Submits a barrier to the queue after all specified events finish.
    pub fn barrier_with_deps(&mut self, dep_events: &[&Event]) -> Result<Event> {
        let dep_events = dep_events
            .iter()
            .map(|e| EventPtr {
                ptr: (*e).clone().0,
            })
            .collect::<Vec<_>>();
        ffi::barrier(&mut self.0, dep_events).map(Into::into)
    }

    /// Performs a blocking wait for the completion of all enqueued tasks in the queue.
    pub fn wait(&mut self) -> Result<()> {
        ffi::wait(&mut self.0)
    }

    /// Enqueues a kernel object to the queue as an ND-range kernel, using the number of work-items
    /// specified by the [`NdRange`] nd_range.
    pub unsafe fn launch<const ARGC: usize, const DIMENSIONS: usize>(
        &mut self,
        nd_range: NdRange<DIMENSIONS>,
        kernel: &Kernel,
        args: impl KernelArgumentList<ARGC>,
    ) -> Result<Event>
    where
        NdRange<DIMENSIONS>: ValidDimension,
    {
        unsafe { nd_range.launch(self, kernel, args) }
    }

    /// Copies the contents of the source buffer to the destination buffer.
    ///
    /// Panics if the source and destination buffer lengths differ.
    pub fn copy<T, A1, A2>(&mut self, src: &Buffer<T, A1>, dst: &mut Buffer<T, A2>) -> Result<Event>
    where
        T: Pod,
        A1: UsmAlloc,
        A2: UsmAlloc,
    {
        self.copy_with_deps(src, dst, &[])
    }

    /// Copies the contents of the source buffer to the destination buffer after all specified
    /// events finish.
    ///
    /// Panics if the source and destination buffer lengths differ.
    pub fn copy_with_deps<T, A1, A2>(
        &mut self,
        src: &Buffer<T, A1>,
        dst: &mut Buffer<T, A2>,
        dep_events: &[&Event],
    ) -> Result<Event>
    where
        T: Pod,
        A1: UsmAlloc,
        A2: UsmAlloc,
    {
        assert_eq!(
            src.get_len(),
            dst.get_len(),
            "source and destination buffer lengths differ"
        );

        // TODO: Resolve the C++ lifetime elision issue
        let dep_events = dep_events
            .iter()
            .map(|e| EventPtr {
                ptr: (*e).clone().0,
            })
            .collect::<Vec<_>>();

        let num_bytes = src.get_len() * size_of::<T>();
        unsafe {
            ffi::memcpy(
                &mut self.0,
                dst.get_byte_ptr(),
                src.get_byte_ptr(),
                num_bytes,
                dep_events,
            )
        }
        .map(Into::into)
    }
}

impl From<&Device> for Queue {
    fn from(value: &Device) -> Self {
        Self(ffi::new_queue_from_device(&value.0))
    }
}

impl From<cxx::UniquePtr<ffi::Queue>> for Queue {
    fn from(value: cxx::UniquePtr<ffi::Queue>) -> Self {
        Self(value)
    }
}

impl Clone for Queue {
    fn clone(&self) -> Self {
        ffi::clone(&self.0).into()
    }
}
